### ISSUE TITLE
Cleanup job folders to reduce number of files

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-server#474]
From https://github.com/fractal-analytics-platform/fractal-server/issues/473
> * Number of files in job folders is currently too large, because each task will have something like 7 files (out/err/in.json/out.json/slurm.sh/slurm.out/slurm.err). If a dataset has 400 wells and a workflow has 5 tasks, this means 14k files, which is not acceptable. To-do: think about how to clean up

Multiple cleanup strategies on the server-side folder:
* combine files (after they are not any more required to provide precise debugging logs)
* tar-zip files
* remove files

Another question is what to do on the (currently being implemented) user directory.

This is a placeholder issue, more on this at a later point.

**Comment by tcompa** [comment]
A reminder for when we address this issue: should we avoid writing the `metadiff` file of *parallel* tasks?

For the moment, the file content is always `{}`, and it is fully ignored by fractal-server (that is, it could be anything and it wouldn't matter). Still, we have to write it N times (where N is the number of parallel components times the number of parallel tasks, e.g. 384*5~2k), and in the SLURM backend we also have to sudo-copy it N times (which is possibly slow, ref #615).

**Comment by tcompa** [comment]
For the record, we now observed at least one case of a job folder taking up ~500M.

Metadata seem to include a reasonably large number of components:
```
>>> import json
>>> with open("1_par_300620234compoundAssaycycle0_zarr_E_12_4.args.json", "r") as f: args = json.load(f)
... 
>>> list(args.keys())
['input_paths', 'output_path', 'metadata', 'overwrite', 'component']
>>> list(args["metadata"].keys())
['plate', 'well', 'image', 'num_levels', 'coarsening_xy', 'original_paths', 'image_extension', 'image_glob_patterns']

>>> len(args["metadata"]["image"])
696

>>> len(args["metadata"]["well"])
116
```
and the number of files is huge
```
$ ls -ls | wc -l
36199
$ ls -l | wc -l
36199
$ ls -l *.json | wc -l
15548
$ ls -l *.err | wc -l
9474
$ ls -l 1_*.err | wc -l
796
$ ls -l 2_*.err | wc -l
796
$ ls -l 12_*.err | wc -l
870
$ ls -l 13_*.err | wc -l
ls: cannot access '13_*.err': No such file or directory
0
```

**Comment by tcompa** [comment]
Here is a real-life situation:
```console
$ cd /some/job/folder

$ $ du -sh
18G	.

$ ls -l | grep json | wc -l
51213

$ ls -l | grep ".json" | wc -l
51213

$ ls -l | grep "\.json" | wc -l
51213

$ ls -l | grep "\.err" | wc -l
38044

$ ls -l | grep "\.out" | wc -l
38044

$ ls -l | grep "\.pickle" | wc -l
5868
```
which suddenly makes this issue much more relevant.

**Comment by jluethi** [comment]
Oh wow!

So this is a multiplexing plate with > 300 wells and 26 multiplexing cycles. Also, the corresponding workflow consists of 18 tasks.

So I would expect something like 8000 different Zarr images in this Zarr plate to be processed here. It's a real test of the scalability of different parts!

(one side observation: The Create OME-Zarr structure multiplexing task also runs for >40 minutes on this dataset! And still going 🤯)

**Comment by jluethi** [comment]
As it relates to logs: Job 1006 is a realistic job that ran fine until there was some cluster maintenance that lead to a job execution error in the 5th task of the workflow.

The pickle files could be from this job execution errors, but the json files likely are the real expected ones. e.g. multiple json files per task per image can scale fast if we have ~8000 images to process.

Thus, I think it's worth looking at the structure of the log files and see where they scale unexpectedly. This is not a side effect of multi-submission or anything like that, but a bigger stress-test than the earlier examples.

**Comment by tcompa** [comment]
Before entering a full analysis, here are a couple of preliminary comments after a quick look.

1. Because of the old handling of empty metadata, at the moment we are writing a lot of empty JSON files. I think we can now simply skip writing them, when they are empty (fractal-tasks-core side). This won't change the total size, but for sure it's best not to have such a large number of file.

2. The biggest contribution to this huge size, luckily, comes from an attribute that is meant to go away in the near future, namely the `dataset.metadata` attribute which is propagated to _all_ tasks. Even when running in parallel, every single instance of a task receives the full `metadata` attribute, which includes the list of all images in the dataset. The results below (numbers are _approximate_ sizes in bytes, obtained as in https://code.activestate.com/recipes/577504/), show that this is what takes the largest part of memory
```
--------------------------------------------------------------------------------
input_paths           : total_size(value)=205
output_path           : total_size(value)=117
metadata              : total_size(value)=1029273
reference_cycle       : total_size(value)=50
overwrite_input       : total_size(value)=28
registered_roi_table  : total_size(value)=74
component             : total_size(value)=98
--------------------------------------------------------------------------------
metadata    plate                 : total_size(value)=178
metadata    well                  : total_size(value)=37171
metadata    image                 : total_size(value)=981738
metadata    num_levels            : total_size(value)=28
metadata    coarsening_xy         : total_size(value)=28
metadata    original_paths        : total_size(value)=9225
metadata    image_extension       : total_size(value)=52
metadata    image_glob_patterns   : total_size(value)=16
--------------------------------------------------------------------------------
```
This shows, once more, why the `metadata` argument of the tasks-API V1 is not meant to remain as is in V2: we are sending some "global" (dataset-wise) information to each and every single-image execution. The second largest contribution comes from `original_paths`, which is another example of a global information being propagated to all parallel tasks. However this only scales with the number of multiplexing cycle, if I'm seeing correctly, and then it's a much smaller issue (see also sizes reported above).


Side-note: we'll see the same scaling (with the number of images) in V2 in the `dataset.images` attribute (and even worse, because there will be multiple versions of the same image). This is a different kind of problem because:
* It's not a disk-space problem
* It only scales linearly with the number of images (vs the current quadratic scaling due to N files having a list of N images each)
* It _will_ matter for all operations we are prototyping in https://github.com/fractal-analytics-platform/tinkering-flexibility-and-task-api, since they'll need to be reasonably efficient

TLDR The switch to task-API V2 would strongly mitigate this issue. I'm not saying that this is the end of it, and we need to both apply minor fixes when relevant (e.g. let's not write empty JSON files) and review the overall questions of what gets written (and what is erased).

**Comment by tcompa** [comment]
Addendum: the results above are for a single args.json file, which has a size of about 600K.

**Comment by tcompa** [comment]
> (one side observation: The Create OME-Zarr structure multiplexing task also runs for >40 minutes on this dataset! And still going 🤯)

We did observe some slow behavior for the metadata parsing in the past, ref
* https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/137
* https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/94

But I cannot say whether this is the actual issue. If we are printing some pre-parsing and post-parsing logs, then this information should be available in the err file corresponding to that SLURM job. Let's open an issue on fractal-tasks-core, if appropriate.

**Comment by jluethi** [comment]
Thanks a lot for this analysis, that's super useful to be aware and can inform decisions in the Task API V2 transition! :)

And sounds very promising that we can improve the scaling a lot with our current V2 plans. Let's make sure to run some (simulated) stress tests with a similar size of wells & multiplexing cycles to compare the scaling in V2 drafts once they progress some more

**Comment by tcompa** [comment]
#1237 represents a quick fix for the specific issue of a dataset with a large number of images.
This issue should be revisited during/after the V2 refactor, for a broader review of the job folders.

**Comment by tcompa** [comment]
Closing in favor of #1521 and #1522
